### PR TITLE
fix Issue 17947 - C++ std::pair::swap mangled incorrectly

### PR DESCRIPTION
**Pull request by WalterBright**
Not ready for review. Work in progress. Supercedes https://github.com/dlang/dmd/pull/6658

### CODE REVIEW

**Comment by ibuclaw** [comment]
Perhaps you should have one commit per fix, with a test case. I spotted at least three changes to mangling on a quick passthru.

**Comment by WalterBright** [comment]
This isn't an incremental fix. It's a redo. Besides, you should point out what those changes are so I can double check them! Perhaps they are unintentional.

File: src/ddmd/cppmangle.d [hunk 1]
```diff
@@ -34,225 +34,267 @@ import ddmd.visitor;
...
 {
     alias visit = super.visit;
-    Objects components;
-    OutBuffer buf;
-    bool is_top_level;
+    Objects components;         // array of components available for substitution
+    OutBuffer* buf;             // append the mangling to buf[]
+    bool is_top_level;          // true if ignore 'const' mangling attribute
     bool components_on;
+    Loc loc;                    // location for use in error messages
```
Comment (ibuclaw) [inline src/ddmd/cppmangle.d]: Adding loc is one change that could be separated.

File: src/ddmd/cppmangle.d [hunk 2]
```diff
@@ -603,139 +709,74 @@ public:
          * e        long double, __float80
          * g        __float128
          * z        ellipsis
+         * Dd       64 bit IEEE 754r decimal floating point
+         * De       128 bit IEEE 754r decimal floating point
+         * Df       32 bit IEEE 754r decimal floating point
+         * Dh       16 bit IEEE 754r half-precision floating point
+         * Di       char32_t
+         * Ds       char16_t
```
Comment (ibuclaw) [inline src/ddmd/cppmangle.d]: Do we need to document types we don't support?

File: src/ddmd/cppmangle.d [hunk 3]
```diff
@@ -746,61 +787,50 @@ public:
         {
             assert(t.basetype && t.basetype.ty == Tsarray);
             assert((cast(TypeSArray)t.basetype).dim);
-            //buf.printf("Dv%llu_", ((TypeSArray *)t.basetype).dim.toInteger());// -- Gnu ABI v.4
-            buf.writestring("U8__vector"); //-- Gnu ABI v.3
+            version (all)
+                buf.printf("Dv%llu_", (cast(TypeSArray *)t.basetype).dim.toInteger()); // -- Gnu ABI v.4
+            else
+                buf.writestring("U8__vector"); //-- Gnu ABI v.3
```
Comment (ibuclaw) [inline src/ddmd/cppmangle.d]: Change in vector abi could be a separate commit.

File: src/ddmd/cppmangle.d [hunk 4]
```diff
@@ -746,61 +787,50 @@ public:
...
     override void visit(TypeReference t)
     {
+        //printf("TypeReference %s\n", t.toChars());
         is_top_level = false;
         if (substitute(t))
             return;
         buf.writeByte('R');
         t.next.accept(this);
-        store(t);
+         append(t);
```
Comment (ibuclaw) [inline src/ddmd/cppmangle.d]: Funny whitespace jerr

File: src/ddmd/cppmangle.d [hunk 5]
```diff
@@ -836,58 +866,30 @@ public:
...
-        argsCppMangle(t.parameters, t.varargs);
+        mangleFunctionParameters(t.parameters, t.varargs);
         buf.writeByte('E');
-        store(t);
-    }
-
-    override void visit(TypeDelegate t)
-    {
-        visit(cast(Type)t);
+         append(t);
```
Comment (ibuclaw) [inline src/ddmd/cppmangle.d]: Looks like this is not aligned either

File: src/ddmd/cppmangle.d [hunk 6]
```diff
@@ -896,65 +898,57 @@ public:
...
-            if (t.isImmutable() || t.isShared())
+            if (!substitute(s))
             {
-                visit(cast(Type)t);
+                cpp_mangle_name(s, t.isConst());
             }
         }
         if (t.isConst())
-            store(t);
+             append(t);
```
Comment (ibuclaw) [inline src/ddmd/cppmangle.d]: Looks like you added an extra space in your automated find/replacement of this function.

File: src/ddmd/cppmangle.d [hunk 7]
```diff
@@ -529,59 +598,96 @@ private final class CppMangleVisitor : Visitor
...
             assert(0);
         }
-        Target.prefixName(&buf, LINKcpp);
-        return buf.extractString();
     }
 
-    override void visit(Type t)
+    /****** The rest is type mangling ************/
+
+    void error(Type t)
```
Comment (ibuclaw) [inline src/ddmd/cppmangle.d]: Refactoring errors could be separate commit.

File: src/ddmd/cppmangle.d [hunk 8]
```diff
@@ -529,59 +598,96 @@ private final class CppMangleVisitor : Visitor
...
+
+    /******
+     * Write out 1 or 2 character basic type mangling.
+     * Handle const and substitutions.
+     * Params:
+     *  t = type to mangle
+     *  p = if not 0, then character prefix
+     *  c = mangling character
+     */
+    void writeBasicType(Type t, char p, char c)
```
Comment (ibuclaw) [inline src/ddmd/cppmangle.d]: Refactoring writeBasicType could be a separate commit.

File: src/ddmd/cppmangle.d [hunk 9]
```diff
@@ -34,225 +34,267 @@ import ddmd.visitor;
...
 extern (C++):
 
-extern (C++) const(char)* toCppMangleItanium(Dsymbol s)
+const(char)* toCppMangleItanium(Dsymbol s)
 {
     //printf("toCppMangleItanium(%s)\n", s.toChars());
-    scope CppMangleVisitor v = new CppMangleVisitor();
-    return v.mangleOf(s);
+    OutBuffer buf;
+    Target.prefixName(&buf, LINKcpp);
```
Comment (ibuclaw) [inline src/ddmd/cppmangle.d]: Moving this call here.  Rather than being inside mangleOf.

File: test/compilable/cppmangle.d [hunk 10]
```diff
@@ -378,3 +378,19 @@ else version (Posix)
...
+{
+    struct pair(T1, T2)
+    {
+	void swap(ref pair other);
+    }
+}
+
+version (linux)
+{
+    static assert(std.pair!(void*, void*).swap.mangleof == "_ZNSt4pairIPvS0_E4swapERS1_");
```
Comment (ibuclaw) [inline test/compilable/cppmangle.d]: Finally, what changes is actually related to this fix? All I can see are refactorings.

**Review by ibuclaw — COMMENTED** [review]
The two Abi changes I can see are __vector and whatever was done to fix the test case.

File: src/ddmd/cppmangle.d (hunk 6 above)
Comment (WalterBright) [reply src/ddmd/cppmangle.d]: indeed I did. oops!

File: src/ddmd/cppmangle.d (hunk 2 above)
Comment (WalterBright) [reply src/ddmd/cppmangle.d]: I'd prefer that wchar and dchar map to them, but that would break backward compatibility. Not sure what to do here, so I documented it.

File: src/ddmd/cppmangle.d (hunk 9 above)
Comment (WalterBright) [reply src/ddmd/cppmangle.d]: I moved it because it should be done first, not last, as doing it last requires rippling the mangled name up in the buffer. It's not an ABI change.

Comment (ibuclaw) [reply src/ddmd/cppmangle.d]: Yep, I did make a mental note that it avoids a realloc/memmove.

**Comment by WalterBright** [comment]
> The two Abi changes I can see are __vector

My Linux machine g++ uses the newer one. There are no test cases for it in the test suite. When I finish the refactorings, I intend to add some test cases to improve coverage testing.

> and whatever was done to fix the test case.

The code is such a mess I'm essentially redesigning/rewriting it, it's not an incremental thing.

**Comment by ibuclaw** [comment]
> My Linux machine g++ uses the newer one. There are no test cases for it in the test suite. When I finish the refactorings, I intend to add some test cases to improve coverage testing.

Just a suggestion then, maybe there could be a new cppAbi flag added to target which would be the baseline C++ Abi we support.  Probably gdc would be the only user of such a flag though (exposing it to the command line).

File: test/compilable/cppmangle.d (hunk 10 above)
Comment (WalterBright) [reply test/compilable/cppmangle.d]: I added a bug report for the fix, and commented the code where the two corrections are.

**Comment by WalterBright** [comment]
> Just a suggestion then, maybe there could be a new cppAbi flag added to target which would be the baseline C++ Abi we support.

I don't really like to pretend to support something we don't. I could be convinced if anyone is still using a system with the old mangling - wasn't this change made many years ago?

**Comment by ibuclaw** [comment]
> I don't really like to pretend to support something we don't. I could be convinced if anyone is still using a system with the old mangling - wasn't this change made many years ago?

The problem is the cross over period.  Where you will get linker errors if you're on the other side of the fence.

For instance, the change from C++98 to C++11 comes to mind. https://issues.dlang.org/show_bug.cgi?id=14956

File: src/ddmd/cppmangle.d (hunk 3 above)
Comment (WalterBright) [reply src/ddmd/cppmangle.d]: Yeah, I'll do that separately. https://issues.dlang.org/show_bug.cgi?id=17948

**Comment by WalterBright** [comment]
> The problem is the cross over period.

I understand. My question is is the crossover period in the past?

File: test/compilable/cppmangle.d (hunk 10 above)
Comment (ibuclaw) [reply test/compilable/cppmangle.d]: Thanks, I see it.

**Comment by WalterBright** [comment]
There's much more I want to do with this, but I'll stop here for this PR, as it resolves the immediate crisis of https://issues.dlang.org/show_bug.cgi?id=17947

**Comment by ibuclaw** [comment]
> I understand. My question is is the crossover period in the past?

According to gcc documentation, ABIv2 was default up until g++-4.9.  ABIv4 was first selectable in g++-4.5.   ABIv8 became the default in g++-5.

Two years is two years, but still not a long time for C++ standards.  There are still LTS systems are still in support that have older g++ versions.  The first Ubuntu LTS with a newer ABI was 16.04.

**Comment by WalterBright** [comment]
I backed out that change for now, leaving it up to a future PR.

File: src/ddmd/cppmangle.d [hunk 11]
```diff
@@ -34,225 +34,267 @@ import ddmd.visitor;
...
+    void template_args(TemplateInstance ti)
     {
-        //printf("source_name(%s)\n", s.toChars());
-        TemplateInstance ti = s.isTemplateInstance();
-        if (ti)
-        {
-            if (!skipname && !substitute(ti.tempdecl))
+        /* <template-args> ::= I <template-arg>+ E
+         */
+        if (!ti)                // could happen if std::basic_string is not a template
```
Comment (andralex) [inline src/ddmd/cppmangle.d]: Love this. Everywhere else the code in this PR has the shape:

```d
T some_function(SomeArguments some_arguments)
{
    if (condition)
    {
        ... many, many, MANY lines of code ...
        return result;
    }
    return corner_case;
}
```

which is distinctly unsavory for some obvious and some less obvious reasons.

**Review by andralex — APPROVED** [review]
@ibuclaw made a good point - please make PRs smaller so they're easier to review. I'll approve this one because we don't want to err on the side of bureaucracy, either. Thx!
